### PR TITLE
Feature database

### PR DESCRIPTION
**Pull request by Pletkininkas**
Implementation of a Repository design pattern(to be able to read and modify data using different providers such as MySQL). Binary write/read modification for object and shopping item list binary save.
@jurgiskg

### CODE REVIEW

**Review by jasdom — APPROVED** [review]

File: CSE.BL/BinaryFileManager.cs [hunk 1]
```diff
@@ -1,31 +1,75 @@
...
         {
-            using (Stream stream = File.Open(filePath, append ? FileMode.Append : FileMode.Create))
+            try
+            {
+                using (Stream stream = File.Open(filePath, FileMode.Create))
+                {
+                    var binaryFormatter = new System.Runtime.Serialization.Formatters.Binary.BinaryFormatter();
+                    binaryFormatter.Serialize(stream, objectToWrite);
+                }
+            } catch (DirectoryNotFoundException)
```
Comment (jurgiskg) [inline CSE.BL/BinaryFileManager.cs:19]: why are both exceptions swallowed?

File: CSE.BL/BinaryFileManager.cs [hunk 2]
```diff
@@ -1,31 +1,75 @@
...
+                        return (List<T>)binaryFormatter.Deserialize(stream);
+                    }
+                }
+                else
+                {
+                    return new List<T>();
+                }
+            }
+            catch (Exception)
+            {
```
Comment (jurgiskg) [inline CSE.BL/BinaryFileManager.cs:47]: the exception handling in this and the following method is interesting - is an empty file and a not existing file really the same scenario from the end user's perspective?

File: CSE.BL/ShoppingList/ShoppingItem.cs [hunk 3]
```diff
@@ -1,21 +1,48 @@
...
 namespace CSE.BL.ShoppingList
 {
     /*ShoppingItem object holds data to describe 
-     * the name, amount, and measurement unit type*/
-
+     * id, shopType name, measurement unit type, amount and price.*/
     [Serializable]
     public class ShoppingItem
     {
+        public Dictionary<string, double> dictionary;//dictionary for saving same item prices over different shops
```
Comment (jurgiskg) [inline CSE.BL/ShoppingList/ShoppingItem.cs:11]: Is this field really responsibility of class ShoppingItem?

File: CSE.BL/ShoppingList/ShoppingItem.cs [hunk 4]
```diff
@@ -1,21 +1,48 @@
...
+                return id;
+            }
+            set
+            {
+                id = value;
+            }
+        }
+
+        // get/set index of shop
+        public ShopTypes Shop
```
Comment (jurgiskg) [inline CSE.BL/ShoppingList/ShoppingItem.cs:32]: Why have these explicit properties with underlying fields when
`public ShopTypes Shop { get; set; }
`does the trick?

File: CSE.BL/ShoppingList/ShoppingItem.cs [hunk 5]
```diff
@@ -58,19 +85,50 @@ public UnitTypes Unit
...
+            dictionary.Add(shop.ToString(), price);
+            this.id = id;
+            this.shop = shop;
             this.name = name;
             this.amount = amount;
             this.unit = unit;
+            this.price = price;
         }
 
         public ShoppingItem(ShoppingItem item)
```
Comment (jurgiskg) [inline CSE.BL/ShoppingList/ShoppingItem.cs:120]: a constructor, which takes itself as an argument, is, amm, interesting.
Why not just use C#'s object initialization in the calling class:
```
var shoppingItem = new ShoppingItem()
{
  Id = id,
  Shop = shop,
  ...
}
```

**Pull request merged**